### DISCUSSION TITLE
How safe is it to share a cache for the same compiler but different targeted CPUs?

### DISCUSSION THREAD

**Discussion by Romain-Geissler-1A** [ccache/ccache#725] [Questions & answers] [answered]
### Question ###

Hi,

I would like to know how safe is it to share the same ccache folder when let's say we build the exact same code with the very same toolchain version (so exactly the same gcc or clang commit) both on x86_84 and aarch64, the only thing that changes between both is the default targetted architecture, which is *NOT* part of the compilation command line. And in both case the compiler is invoked with the name "gcc" (or "clang"), there is no string on the command line which can discriminate one architecture over the other ? I expect that in a case where a glibc header is included, most likely the preprocessor output is different, but let's say I use ccache for a very basic C file like containing just `int main(){}` where the preprocessor output doesn't change, is it still ok ? Or shall I really try to ensure we do use explicitly either a different ccache folder, either a different command line ?

Note: actually with qemu, the gcc compiler for aarch64 can even be lauched from the very same x86_64 machine, in case you expect to issue some "uname" syscall to find the machine name (I don't know if qemu rewrites the uname machine name though).

Cheers,
Romain

**Comment by jrosdahl** [comment, marked as answer]
By default, ccache includes the compiler's basename, its modification time ("mtime") and its size in the hash to associate the result with the compiler. If the two compilers (for different architectures) are built from the same source code they likely still will have different mtimes/sizes. If that's not the case then yes, there is a risk for collisions and you'll have to configure another method of identifying the compiler using the [`compiler_check`](https://ccache.dev/manual/4.1.html#config_compiler_check) option, or, as you say, use different cache directories.

**Comment by Romain-Geissler-1A** [comment]
Thanks for this answer. With mtime/size yes it should be pretty safe, so I will keep using the same ccache directory.
